docs/interpreter.md

```diff
@@ -12,9 +12,12 @@ representing its data laid out in
 objects are reference-counted to simplify memory management.
 
 Individual elements of a tensor are represented using `Element` class which uses
-`mlir::Attribute` for storage. Using `mlir::Attribute` simplifies things because
-this means that we don't have to implement our own machinery for storing values
-of different types inside `Element`.
+discriminated union holding one of `APInt` or `vector<APFloat>` for storage. The
+later is used for storing elements with floating-point or complex types. Note
+that we use `mlir::Attribute` while constructing an `Element` object (via
+    `Element(Type type, Attribute attr)`) or extracting value from `Element`
+object (via `Attribute Element::getValue()`), in order to simplify these
+interfaces.
```

Comment:
To be honest, I'm not sure what we're getting out of replacing the storage with `std::variant` if we're still using `mlir::Attribute` as the API.

I think that we should go all the way and completely remove all uses of `Attribute` in this file:
  1) Replace `Element::Element(Type type, Attribute attr)` with three constructors: `Type+APInt`, `Type+APFloat`, `Type+std::complex<APFloat>`.
  2) Replace `Element::getValue` with three methods: `getIntValue`, `getFloatValue` and `getComplexValue` (which are already defined in Element.cpp anyway).
  3) Update `<<` to work without attributes.

This will significantly simplify existing code. E.g. we'll be able to replace swaths of code like `return Element(elementType, FloatAttr::get(elementType, value)` with `return Element(elementType, value)`.

Reply:
I agree with the point that using Attributes in Element's interface API incurs huge overhead.
DOne.